internal/provider/kubernetes/gatewayclass.go

```diff
@@ -103,6 +103,17 @@ func (r *gatewayClassReconciler) Reconcile(ctx context.Context, request reconcil
 
 	for i := range gatewayClasses.Items {
 		if gatewayClasses.Items[i].Spec.ControllerName == r.controller {
+			// The gatewayclass was marked for deletion and the finalizer removed,
+			// so clean-up dependents.
+			if !gatewayClasses.Items[i].DeletionTimestamp.IsZero() &&
+				!slice.ContainsString(gatewayClasses.Items[i].Finalizers, gatewayClassFinalizer) {
+				r.log.Info("gatewayclass marked for deletion")
+				cc.removeMatch(&gatewayClasses.Items[i])
+				// A nil gatewayclass removes managed proxy infra, if it exists.
+				r.resources.GatewayClasses.Store(request.Name, nil)
```

Comment:
this line will need to be revisited either as part of https://github.com/envoyproxy/gateway/issues/423 or post that